src/ds++/VectorView.hh

```diff
@@ -0,0 +1,138 @@
+   */
+  template <typename... Indices> size_t idx(Indices... indices) {
+    Insist(sizeof...(indices) == dim,
+           std::string("VectorView expected ") + std::to_string(dim) +
+               " arguments but only found " +
+               std::to_string(sizeof...(Indices)) + ".");
+    size_t idx = 0, i = 0;
+    using untilFold = int[];
+    // In the following line, the 'pack' is indices and the operation is the
+    // comma. For an explanation see
+    // https://www.codingame.com/playgrounds/2205/7-features-of-c17-that-will-simplify-your-code/fold-expressions
+    // Note that we don't use the computed value 'untilFold', but its
+    // construction has the side effect of setting idx.
+    (void)untilFold{((idx += indices * strides[i++]), 0)...};
+    Check(idx < data.size());
+    return idx;
+  }
+
+  //--------------------------------------------------------------------------//
+  /*! \brief Return the data associated with the provided multi-D view.
+   *
+   * \tparam Indices variatic template that represents a set of integral values,
+   *         one for each dimension of the 'view'.
+   * \param indices a list of integral types that indicate the multi-D offset.
+   * \return Data from the 1-D std::vector<T> associated with the provided
+   *         multi-D index set.
+   */
+  template <typename... Indices> T operator()(Indices... indices) {
+    return data[idx(indices...)];
+  }
+
+  //--------------------------------------------------------------------------//
+  /*! \brief Default constructor
+   *
+   * \param[in] data a reference to the underlying std::vector<T> that we are
+   *               creating a 'view' for.
+   *
+   * When the constructor is called, we must setup the 'stride' or offset values
+   * for each of indices that will be used for this view.
+   */
+  VectorView(std::vector<T> const &data_) : data(data_) {
+    Ensure(dim > 0);
+    // Member data 'strides' was initialized with Extents, but must now be
+    // replaced with actual offset values.
+    for (size_t i = dim - 1; i < dim; --i) {
+      Insist(strides[i] > 0, "Each dimension must have a positive length.");
+      strides[i] = 1;
+      for (size_t j = 0; j < i; j++) {
+        strides[i] *= strides[j];
+      }
+    }
+  }
+
+private:
+  //
+  // Disabled member functions to ensure correct behavior.
+  //
+
+  // Ensure that the idx member function is not defined for the template case
+  // of zero arguments.
+  size_t idx(void) = delete;
+  // Ensure that the paren-operator member function is not defined for the
+  // template case of zero arguments.
+  T operator()() = delete;
+
+  //
+  // >> MEMBER DATA (state)
+  //
+
+  //! The actual data container that we are creating a view for.
+  std::vector<T> const &data;
+  //! Max size of each dimension for the view.
+  size_t const dim = sizeof...(Extents);
+  //! Offsets associated with each of the view's dimensions.
+  std::array<size_t, sizeof...(Extents)> strides = {Extents...};
```

Comment:
If one does want to avoid STL functions, this would probably want to become `size_t *strides` and have matching `new[]`/`delete[]` calls in constructors/destructors.

Reply:
std::array can be optimized at compile-time, I believe? This should have no overhead?